src/realm/sync/subscriptions.hpp

```diff
@@ -0,0 +1,287 @@
+
+        iterator(const SubscriptionSet* parent, LnkLst::iterator it);
+
+    private:
+        const SubscriptionSet* m_parent;
+        LnkLst::iterator m_sub_it;
+        mutable Subscription m_cur_sub;
+    };
+
+    using const_iterator = const iterator;
+
+    // This will make a copy of this subscription set with the next available version number and return it as
+    // a mutable SubscriptionSet to be updated. The new SubscriptionSet's state will be Uncommitted. This
+    // subscription set will be unchanged.
+    SubscriptionSet make_mutable_copy() const;
+
+    // The query version number used in the sync wire protocol to identify this subscription set to the server.
+    int64_t version() const;
+
+    // The current state of this subscription set
+    State state() const;
+
+    // The error string for this subscription set if any.
+    StringData error_str() const;
+
+    // Returns the number of subscriptions in the set.
+    size_t size() const;
+
+    // A const_iterator interface for finding/working with individual subscriptions.
+    const_iterator begin() const;
+    const_iterator end() const;
+
+    // Returns a const_iterator to the query matching either the name or Query object, or end() if no such
+    // subscription exists.
+    const_iterator find(StringData name) const;
+    const_iterator find(const Query& query) const;
+
+    // A mutable iterator interface for working with individual subscriptions in when the set is in the
+    // Uncommitted state.
+    iterator begin();
+    iterator end();
+
+    // Erases a subscription pointed to by an iterator. Returns the "next" iterator in the set - to provide
+    // STL compatibility. The SubscriptionSet must be in the Uncommitted state to call this - otherwise
+    // this will throw.
+    iterator erase(iterator it);
+
+    // Erases all subscriptions in the subscription set.
+    void clear();
+
+    // Inserts a new subscription into the set if one does not exist already - returns an iterator to the
+    // subscription and a bool that is true if a new subscription was actually created. The SubscriptionSet
+    // must be in the Uncommitted state to call this - otherwise this will throw.
+    //
+    // The Query portion of the subscription is mutable, however the name portion is immutable after the
+    // subscription is inserted.
+    //
+    // If insert is called twice for the same query, this is a no-op.
+    std::pair<iterator, bool> insert(const Query& query, util::Optional<std::string> name = util::none);
+
+    // Updates the state of the transaction and optionally updates its error information.
+    //
+    // You may only set an error_str when the State is State::Error.
+    //
+    // If set to State::Complete, this will erase all subscription sets with a version less than this one's.
+    //
+    // This should be called internally within the sync client.
+    void update_state(State state, util::Optional<std::string> error_str = util::none);
```

Comment:
If this is not supposed to be called by SDKs, can we make this protected? Or is the plan to create proxies for these classes at the OS layer?

Reply:
Let me think about that.

Reply:
I think I will write a PIMPL for them at the OS layer, but I won't do that in this PR - I'll do that when I've got things further integrated and am sure this API is pretty stable/okay.